trace_decoder/src/core.rs

```diff
@@ -533,10 +547,147 @@ fn middle<StateTrieT: StateTrie + Clone>(
+    trim_state: &mut BTreeSet<TrieKey>,
+    state_trie: &mut StateTrieT,
+) -> anyhow::Result<()> {
+    // Ethereum mainnet: EIP-4788
+    #[cfg(feature = "eth_mainnet")]
+    do_beacon_hook(
+        block.block_timestamp,
+        storage,
+        trim_storage,
+        block.parent_beacon_block_root,
+        trim_state,
+        state_trie,
+    )?;
+
+    #[cfg(feature = "cdk_erigon")]
+    do_scalable_hook(
+        block,
+        ger_data,
+        storage,
+        trim_storage,
+        trim_state,
+        state_trie,
+    )?;
+
+    Ok(())
+}
+
+#[cfg(feature = "cdk_erigon")]
+/// Updates the storage of the Scalable and GER contracts, according to
+/// <https://docs.polygon.technology/zkEVM/architecture/proving-system/processing-l2-blocks/#etrog-upgrade-fork-id-6>.
+///
+/// This is Polygon-CDK-specific, and runs at the start of the block,
+/// before any transactions (as per the Etrog specification).
+fn do_scalable_hook<StateTrieT: StateTrie + Clone>(
+    block: &BlockMetadata,
+    ger_data: Option<(H256, H256)>,
+    storage: &mut BTreeMap<H256, StorageTrie>,
+    trim_storage: &mut BTreeMap<ethereum_types::H160, BTreeSet<TrieKey>>,
+    trim_state: &mut BTreeSet<TrieKey>,
+    state_trie: &mut StateTrieT,
+) -> anyhow::Result<()> {
+    use evm_arithmetization::testing_utils::{
+        ADDRESS_SCALABLE_L2, ADDRESS_SCALABLE_L2_ADDRESS_HASHED, GLOBAL_EXIT_ROOT_ADDRESS,
+        GLOBAL_EXIT_ROOT_ADDRESS_HASHED, GLOBAL_EXIT_ROOT_STORAGE_POS, LAST_BLOCK_STORAGE_POS,
+        STATE_ROOT_STORAGE_POS, TIMESTAMP_STORAGE_POS,
+    };
+
+    if block.block_number.is_zero() {
+        return Err(anyhow!("Attempted to prove the Genesis block!"));
+    }
+    let scalable_storage = storage
+        .get_mut(&ADDRESS_SCALABLE_L2_ADDRESS_HASHED)
+        .context("missing scalable contract storage trie")?;
+    let scalable_trim = trim_storage.entry(ADDRESS_SCALABLE_L2).or_default();
+
+    let timestamp_slot_key = TrieKey::from_slot_position(U256::from(TIMESTAMP_STORAGE_POS.1));
+
+    let timestamp = scalable_storage
+        .get(&timestamp_slot_key)
+        .map(rlp::decode::<U256>)
+        .unwrap_or(Ok(0.into()))?;
+    let timestamp = core::cmp::max(timestamp, block.block_timestamp);
+
+    // Store block number and largest timestamp
+
+    for (ix, u) in [
+        (U256::from(LAST_BLOCK_STORAGE_POS.1), block.block_number),
+        (U256::from(TIMESTAMP_STORAGE_POS.1), timestamp),
+    ] {
+        let slot = TrieKey::from_slot_position(ix);
+        scalable_trim.insert(slot);
+
+        // These values are never 0.
+        scalable_storage.insert(slot, alloy::rlp::encode(u.compat()))?;
+        scalable_trim.insert(slot);
+    }
+
+    // Store previous block root hash
+
+    let prev_block_root_hash = state_trie.root();
+    let mut arr = [0; 64];
+    (block.block_number - 1).to_big_endian(&mut arr[0..32]);
+    U256::from(STATE_ROOT_STORAGE_POS.1).to_big_endian(&mut arr[32..64]);
```

Comment:
In the link, the slot number seems to be `1|blockNumber`. But this implementation seems to differ? Or am I missing something?

Reply:
Hmm I'm not sure. In `cdk_erigon`, we call:

```golang
		//save prev block hash
		sdb.scalableSetBlockHash(blockNumber-1, prevBlockHash)
```

with function definition:

```golang
func (sdb *IntraBlockState) scalableSetBlockHash(blockNum uint64, blockHash *libcommon.Hash) {
	// create mapping with keccak256(blockNum,position) -> smt root
	d1 := common.LeftPadBytes(uint256.NewInt(blockNum).Bytes(), 32)
	d2 := common.LeftPadBytes(STATE_ROOT_STORAGE_POS.Bytes(), 32)
	mapKey := keccak256.Hash(d1, d2)
	mkh := libcommon.BytesToHash(mapKey)

	hashAsBigU := uint256.NewInt(0).SetBytes(blockHash.Bytes())

	sdb.SetState(ADDRESS_SCALABLE_L2, &mkh, *hashAsBigU)
}
```

i.e. this line: `mapKey := keccak256.Hash(d1, d2)` is hashing the concatenation `blockNum || STATE_ROOT_STORAGE_POS` i.e. with initial inputs: `blockNumber-1 || 0x01`.
Unless I am also missing something